util/deephash/deephash.go

```diff
@@ -164,26 +164,144 @@ func (h *hasher) hashUint64(i uint64) {
+	if f, ok := typeInfoMap.Load(t); ok {
+		return f.(*typeInfo)
 	}
+	typeInfoMapPopulate.Lock()
+	defer typeInfoMapPopulate.Unlock()
+	return getTypeInfoLocked(t)
+}
 
-	w := h.bw
+func getTypeInfoLocked(t reflect.Type) *typeInfo {
+	if f, ok := typeInfoMap.Load(t); ok {
+		return f.(*typeInfo)
+	}
+	ti := &typeInfo{
+		rtype:       t,
+		isRecursive: typeIsRecursive(t),
+	}
+	typeInfoMap.Store(t, ti)
 
-	if v.CanInterface() {
-		// Use AppendTo methods, if available and cheap.
-		if v.CanAddr() && v.Type().Implements(appenderToType) {
-			a := v.Addr().Interface().(appenderTo)
-			size := h.scratch[:8]
-			record := a.AppendTo(size)
-			binary.LittleEndian.PutUint64(record, uint64(len(record)-len(size)))
-			w.Write(record)
-			return
+	switch t.Kind() {
+	case reflect.Map:
+		ti.keyTypeInfo = getTypeInfoLocked(t.Key())
+		fallthrough
+	case reflect.Ptr, reflect.Slice, reflect.Array:
+		ti.eleTypeInfo = getTypeInfoLocked(t.Elem())
+	}
+	return ti
+}
+
+// typeIsRecursive reports whether t has a path back to itself.
+//
+// For interfaces, it currently always reports true.
+func typeIsRecursive(t reflect.Type) bool {
+	inStack := map[reflect.Type]bool{}
+
+	var stack []reflect.Type
+
+	var visitType func(t reflect.Type) (isRecursiveSoFar bool)
+	visitType = func(t reflect.Type) (isRecursiveSoFar bool) {
+		switch t.Kind() {
+		case reflect.Bool,
+			reflect.Int,
+			reflect.Int8,
+			reflect.Int16,
+			reflect.Int32,
+			reflect.Int64,
+			reflect.Uint,
+			reflect.Uint8,
+			reflect.Uint16,
+			reflect.Uint32,
+			reflect.Uint64,
+			reflect.Uintptr,
+			reflect.Float32,
+			reflect.Float64,
+			reflect.Complex64,
+			reflect.Complex128,
+			reflect.String,
+			reflect.Func:
+			return false
+		}
+		if t.Size() == 0 {
+			return false
+		}
+		if inStack[t] {
+			return true
 		}
+		stack = append(stack, t)
+		inStack[t] = true
+		defer func() {
+			delete(inStack, t)
+			stack = stack[:len(stack)-1]
+		}()
+
+		switch t.Kind() {
+		default:
+			panic("unhandled kind " + t.Kind().String())
+		case reflect.Interface:
+			// Assume the worst for now. TODO(bradfitz): in some cases
+			// we should be able to prove that it's not recursive. Not worth
+			// it for now.
+			return true
+		case reflect.Array, reflect.Chan, reflect.Pointer, reflect.Slice:
+			return visitType(t.Elem())
+		case reflect.Map:
+			if visitType(t.Key()) {
+				return true
+			}
+			if visitType(t.Elem()) {
+				return true
+			}
+		case reflect.Struct:
+			if t.String() == "intern.Value" {
+				// Otherwise its interface{} makes this return true.
+				return false
+			}
+			if t.Size() == 0 {
+				return false
+			}
+			for i, numField := 0, t.NumField(); i < numField; i++ {
+				if visitType(t.Field(i).Type) {
+					return true
+				}
+			}
+			return false
+		}
+		return false
+	}
+	return visitType(t)
+}
+
+func (h *hasher) hashValue(v reflect.Value, forceCycleChecking bool) {
+	if !v.IsValid() {
+		return
 	}
+	ti := getTypeInfo(v.Type())
```

Comment:
There's a race here with concurrent hashes of the same type where `ti.eleTypeInfo` and `ti.keyTypeInfo` are `nil` when first stored into the map, and populated asynchronously.

Perhaps we should do:
```go
type typeInfo struct {
	done uint32
	wait chan struct{} // closed when done is atomically set to 1

	...
}

func getTypeInfoLocked(t reflect.Type) *typeInfo {
	...
	ti := &typeInfo{wait: make(chan struct{}), ...}
	typeInfoMap.Store(t, ti)
	defer func() {
		close(ti.wait)
		atomic.StoreUint32(&ti.done, 1)
	}
	...
}


func (h *hasher) hashValue(v reflect.Value, forceCycleChecking bool) {
	...
	ti := getTypeInfo(v.Type())
	if atomic.LoadUint32(&ti.done) == 0 {
		<-ti.wait
	}
	...
}
```

Reply:
Semantically, we could rely on `<-ti.wait` alone, but apparently an atomic load is 7x faster for the common case.

Reply:
PTAL